### ISSUE TITLE
Migrate away from gogoprotobuf to protobuf v2 API + vtprotobuf

### ISSUE THREAD

**Issue by GiedriusS** [thanos-io/thanos#4557]
### Is your proposal related to a problem?

gogoprotobuf is unmaintained: https://github.com/gogo/protobuf/issues/691

So, we should move to the new protobuf v2 API. Also, we should try to use https://vitess.io/blog/2021-06-03-a-new-protobuf-generator-for-go/ to not make Thanos slower

### Describe the solution you'd like
 
Move to the protobuf V2 API: https://blog.golang.org/protobuf-apiv2. Also, use vtprotobuf so that on Query side we could get all `SeriesResponse`s from a `sync.Pool`

### Describe alternatives you've considered

Alternative is to wait if gogoprotobuf will get a new maintainer and maybe then we wouldn't have to move but that probably will never happen.

### Additional context

N/A

**Comment by yeya24** [comment]
Love this idea!
Can you explain the sentence below? Is this something done automatically by vtprotobuf?

> Also, use vtprotobuf so that on Query side we could get all SeriesResponses from a sync.Pool

**Comment by squat** [comment]
I think he means that by switching to this proto generator we can benefit from using vtproto's memory pooling to greatly reduce the number of allocations needed for umarshaling when receiving message streams during querying, which could otherwise be allocation-heavy, and thus speed up the code path.

**Comment by kakkoyun** [comment]
Yes, this is a great idea! We had a quick chat about this with @bwplotka as well. We should definitely migrate to `vtprotobuf`. I'd be happy to help with this if @GiedriusS you're not already on it.

**Comment by GiedriusS** [comment]
> I think he means that by switching to this proto generator we can benefit from using vtproto's memory pooling to greatly reduce the number of allocations needed for umarshaling when receiving message streams during querying, which could otherwise be allocation-heavy, and thus speed up the code path.

Yep, there's more info here: https://github.com/planetscale/vtprotobuf#available-features. We could call `func YourProtoFromVTPool() *YourProto` to get a `SeriesResponse` on the Querier side that comes from a `sync.Pool` instead of doing this: https://github.com/thanos-io/thanos/blob/main/pkg/store/storepb/rpc.pb.go#L755

It's quite a huge work, I haven't started anything :smile: I thought perhaps this could be a GSoC/LFX project

**Comment by squat** [comment]
> It's quite a huge work, I haven't started anything 😄 I thought perhaps this could be a GSoC/LFX project

Yes i was thinking exactly the same. Shall we propose it? Id be happy to co-mentor

**Comment by iamrajiv** [comment]
Hi, @GiedriusS @squat can I get the opportunity to work on this? I have contributed to gRPC-Gateway and have a decent knowledge of gRPC and Protocol Buffers. I guess we would be referring to [https://pkg.go.dev/google.golang.org/protobuf](https://pkg.go.dev/google.golang.org/protobuf) for APIv2.

**Comment by squat** [comment]
Hi @iamrajiv we just submitted this project to the Linux foundation student mentorship program for the fall https://github.com/cncf/mentoring/blob/main/lfx-mentorship/2021/03-Fall/project_ideas.md#migrate-thanos-to-the-new-protocol-buffers-v2-api
Maybe we can figure out a way multiple people contributing to this though.

**Comment by pkj-m** [comment]
Hi folks! This looks like an interesting project, and in sync with my skillset. I would be happy to work on this and contribute to Thanos as part of the LFX Mentorship program.

I have prior experience with open-source projects through participating (and eventually mentoring) in GSoC. I'm quite proficient in Golang, and have worked with protobuf in the past during my summer internship at Uber, where I was tasked with writing a new proto service from scratch in Go.

I'll take a look at the repository and reach out soon with further questions. Thanks!

**Comment by GiedriusS** [comment]
> Hi folks! This looks like an interesting project, and in sync with my skillset. I would be happy to work on this and contribute to Thanos as part of the LFX Mentorship program.
> 
> I have prior experience with open-source projects through participating (and eventually mentoring) in GSoC. I'm quite proficient in Golang, and have worked with protobuf in the past during my summer internship at Uber, where I was tasked with writing a new proto service from scratch in Go.
> 
> I'll take a look at the repository and reach out soon with further questions. Thanks!

Feel free to apply via the LFX website!

**Comment by yeya24** [comment]
Still valid

**Comment by GiedriusS** [comment]
Our work is here:
https://github.com/rahulii/thanos/pull/1
https://github.com/rahulii/thanos/pull/2

Need to continue with this work :/

**Comment by iamrajiv** [comment]
Hi @GiedriusS is there any doc that updates about how much work has to be done and how much has to be done? I am interested in working on this issue.

**Comment by GiedriusS** [comment]
> Hi @GiedriusS is there any doc that updates about how much work has to be done and how much has to be done? I am interested in working on this issue.

I'm afraid that the PRs are the only place you can find info. Also, I wrote a small post detailing the learnings from doing this: https://giedrius.blog/2022/02/04/things-learned-from-trying-to-migrate-to-protobuf-v2-api-from-gogoprotobuf-so-far/.

**Comment by iamrajiv** [comment]
@GiedriusS Great. I want to continue this work in GSoC 2022.

**Comment by GiedriusS** [comment]
@kakkoyun I am continuing work on this - already switched to upstream protobuf compiler. Just need to fix tests & add pooling now. My branch is available on https://github.com/GiedriusS/thanos/tree/gogo_remove.

**Comment by matej-g** [comment]
Linking the current WIP for more visibility - https://github.com/thanos-io/thanos/pull/5421

**Comment by GiedriusS** [comment]
I did this but it's not worth it: https://github.com/thanos-io/thanos/pull/7790. A better solution would be to focus on grpc/protobuf alternatives. See the PR for more.
